clusterloader2/pkg/measurement/common/network/network_performance_measurement.go

```diff
@@ -0,0 +1,397 @@
+	if err := npm.framework.DeleteObject(crdGvk, "", customResourceDefinitionName); err != nil {
+		klog.Errorf("error while deleting CRD: %v", err)
+	}
+	if err := npm.framework.DeleteObject(rbacGvk, "", rbacName); err != nil {
+		klog.Errorf("error while deleting clusterRoleBinding: %v", err)
+	}
+	if err := client.DeleteNamespace(npm.k8sClient, netperfNamespace); err != nil {
+		klog.Errorf("error while deleting %s namespace: %v", netperfNamespace, err)
+	}
+	if err := client.WaitForDeleteNamespace(npm.k8sClient, netperfNamespace); err != nil {
+		klog.Errorf("error while waiting for %s namespace to be deleted: %v", netperfNamespace, err)
+	}
+}
+
+func (npm *networkPerformanceMeasurement) createAndWaitForWorkerPods() error {
+	// Create worker pods
+	var replicas = map[string]interface{}{"Replicas": npm.numberOfClients + npm.numberOfServers}
+	if err := npm.framework.ApplyTemplatedManifests(workerPodDeploymentManifestFilePath, replicas); err != nil {
+		return fmt.Errorf("error while creating worker pods :%v ", err)
+	}
+	// Wait for all worker pods to be ready
+	stopCh := make(chan struct{})
+	time.AfterFunc(podReadyTimeout, func() {
+		close(stopCh)
+	})
+	options := &measurementutil.WaitForPodOptions{
+		Selector:            &measurementutil.ObjectSelector{Namespace: netperfNamespace},
+		DesiredPodCount:     npm.numberOfClients + npm.numberOfServers,
+		CallerName:          networkPerformanceMetricsName,
+		WaitForPodsInterval: 2 * time.Second,
+	}
+	return measurementutil.WaitForPods(npm.k8sClient, stopCh, options)
+}
+
+func (*networkPerformanceMeasurement) String() string {
+	return networkPerformanceMetricsName
+}
+
+func newWorkerPodData(podName, nodeName, podIP string) *workerPodData {
+	return &workerPodData{podName: podName, workerNode: nodeName, podIP: podIP}
+}
+
+func (npm *networkPerformanceMeasurement) storeWorkerPods() error {
+	pods, err := npm.k8sClient.CoreV1().Pods(netperfNamespace).List(context.TODO(), metav1.ListOptions{})
+	if err != nil {
+		return err
+	}
+	for _, pod := range pods.Items {
+		if pod.Status.PodIP == "" {
+			klog.Errorf("IP address not found for %s", pod.Name)
+			continue
+		}
+		npm.addPodWorker(newWorkerPodData(pod.Name, pod.Spec.NodeName, pod.Status.PodIP))
+	}
+	return nil
+}
+
+func (npm *networkPerformanceMeasurement) gather() (measurement.Summary, error) {
+	close(npm.stopCh)
+	npm.cleanupCluster()
+	resultSummary := npm.createResultSummary()
+	content, err := util.PrettyPrintJSON(&measurementutil.PerfData{
+		Version:   "v1",
+		DataItems: resultSummary.DataItems,
+	})
+	if err != nil {
+		klog.Info("Pretty Print to Json Err: ", err)
+	}
+	summaryName := fmt.Sprintf(npm.String() + resultSummary.podRatio + resultSummary.protocol + resultSummary.service)
+	return measurement.CreateSummary(summaryName, "json", content), nil
+}
+
+// Dispose disposes resources,objects after the measurement.
+func (npm *networkPerformanceMeasurement) Dispose() {
+	if npm.framework == nil {
```

Comment:
This if should be part of cleanupCluster (in theory nothing is protecting you from calling gather before start, which would lead to crash.

Reply:
modified